### PR TITLE
test: Add test for rpc_whitelist

### PR DESCRIPTION
**Pull request by cvengler**
A test for #12763
Please only review `test/function/rpc_whitelist.py`. Other stuff should belong to the original PR

### CODE REVIEW

**Comment by JeremyRubin** [comment]
The test looks good so far!

Some other conditions to check:

```
rpcwhitelist=user1:A,B,C
rpcwhitelist=user1:B
```

results in only B being available, not A or C.


Test empty rpcwhitelist
```
rpcwhitelist=user1
rpcwhitelist=user2:
```

Test trailing comma
```
rpcwhitelist=user1:A,
```


Test the above where `whitelistrpcdefault=1` is set. This case is really difficult to do because the testing framework assumes access to the RPCs for setup. So you can set up setting the args for the test-framework's rpc credential to be sufficient, but then testing the case where there are no credentialed users seems... hard! But maybe we're OK with not covering that case via test.

It also looks like there is a new method (since I wrote the original test sketch?) called "add_options", which you could override to add the command line args... but what's written works so if it ain't broke...

**Comment by emilengler** [comment]
@JeremyRubin 
The test was written from scratch with a method taken from `rpc_users.py`
I will add your suggestions now

**Comment by JeremyRubin** [comment]
Ah yeah -- I just mean that I think the add_options method appears to be new -- newer than rpc_users too I guess? Not sure!

**Comment by JeremyRubin** [comment]
BTW (maybe a minority opinion here) but those linters are *super* pesky.

You need to fix the typos on "permissions" in order for the tests to run.

**Comment by emilengler** [comment]
@JeremyRubin Ok, will change that.
I also added the exceptions mentioned in your first comment. See f012f128c7c68a4f743cb06f7616997653943352

**Comment by emilengler** [comment]
BTW where you've found the type you meant in the comment? Any line number please?

**Comment by MarcoFalke** [comment]
The linters fail because trailing white space was added. Trailing white space is causing issues with `git` and many editors. The type can be left as is if you want to.

**Comment by emilengler** [comment]
Oh now I understand what you mean :D
Should be fixed now

**Comment by fanquake** [comment]
@emilengler You need need to [squash your commits](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#squashing-commits). If you're fixing things like typos and whitespace, you'll always need to squash, so it's best to just do it before pushing to GitHub.

**Comment by emilengler** [comment]
@fanquake Done

File: test/functional/rpc_whitelist.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    def run_test(self):
+        for user in self.users:
+            permissions = user[2].replace(" ", "").split(",")
+            # Pop all empty items
+            i = 0
+            while i < len(permissions):
+                if permissions[i] == '':
+                    permissions.pop(i)
+
+                i += 1
```
Comment (JeremyRubin) [inline test/functional/rpc_whitelist.py]: just use `import re; filter(lambda x: x=='', re.split(" |,", user[2]))`?

File: test/functional/rpc_whitelist.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+            self.log.info("[" + user[0] + "]: Testing a permitted permission")
+            assert_equal(200, rpccall(self.nodes[0], user, permissions[randint(0, len(permissions) - 1)]).status)
```
Comment (JeremyRubin) [inline test/functional/rpc_whitelist.py]: I like what you're thinking here with picking a random permission (especially if we eventually make this scale out to test a larger set), but I think for functional tests we value the determinism & repeatability a bit more. So maybe as a compromise, test everything in permissions, and then test all of the disallowed (until in the future they are too many).

I'd modify this to make it test all of the permissions

**Review by JeremyRubin — CHANGES_REQUESTED** [review]
Looking better! Still need to test `rpcwhitelistdefault=1`. Tagging @jnewbery who might have some ideas on how to do that?

Maybe one more test to add:

rpcwhitelist=user1:A,A

Just make sure that A is permitted/doesn't crash when named twice

**Comment by MarcoFalke** [comment]
It might be possible to adjust `def wait_for_rpc_connection` to accept the 403 error code as acceptable signal that indicates the rpc thread is running.

File: test/functional/rpc_whitelist.py (hunk 1 above)
Comment (emilengler) [reply test/functional/rpc_whitelist.py]: ACK

File: test/functional/rpc_whitelist.py (hunk 2 above)
Comment (emilengler) [reply test/functional/rpc_whitelist.py]: Ok

Comment (emilengler) [reply test/functional/rpc_whitelist.py]: Done

Comment (JeremyRubin) [reply test/functional/rpc_whitelist.py]: I think you left a few instances lower down where you still do random :)

Comment (emilengler) [reply test/functional/rpc_whitelist.py]: Oh right, forgot the strange tests

Comment (emilengler) [reply test/functional/rpc_whitelist.py]: Done

**Comment by emilengler** [comment]
The test is done.
The only problem is to test `rpcwhitelistdefault=1`
There are two possible ways:

- The easy way: Just add another test
- The hard way: Restart bitcoind with other args during the test. Any ideas on how to do this @MarcoFalke?

**Comment by JeremyRubin** [comment]
You don't need a new file, you can just do a new subclass of the test framework in that file and call one after the other in Main.

It will be a bit tough because you can't actually stop a node using RPC with rpcwhitelistdefault=1 and no auth with stop permitted. So you'd need to write a bit of code to kill by PID.

**Comment by emilengler** [comment]
> You don't need a new file, you can just do a new subclass of the test framework in that file and call one after the other in Main.

Tried that and failed, will try it again tomorrow

**Comment by JeremyRubin** [comment]
@emilengler checking in here -- stuck?

**Comment by emilengler** [comment]
@JeremyRubin The test works fine. The only problem is testing `rpcwhitelistenable=1`. The problem is that I haven't found a way to restart bitcoind. It would need to be separate test.

**Comment by JeremyRubin** [comment]
I'm OK with that; great work preparing these tests.

@MarcoFalke given the strong concept ACKs+utacks/lgtms on the original PR, now with @emilengler's work on tests I think this is in a mergeable state. Do you prefer for me to squash the existing branch (I don't want to make people have to re-ack the squash needlessly)?

**Comment by emilengler** [comment]
> @emilengler's work on tests I think this is in a mergeable state. Do you prefer for me to squash the existing branch (I don't want to make people have to re-ack the squash needlessly)?

I'm a bit afraid that mine will cause merge issues then. Otherwise I will rebase it from your updated branch.

**Comment by JeremyRubin** [comment]
@emilengler you can rebase

**Comment by emilengler** [comment]
@JeremyRubin Should be done now

**Comment by emilengler** [comment]
Got merged into master by `2081442c421cc4376e5d7839f68fbe7630e89103`

**Pull request closed without merging**